// @onlyworlds/sdk/dist/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  ELEMENT_ICONS: () => ELEMENT_ICONS,
  ELEMENT_LABELS: () => ELEMENT_LABELS,
  ELEMENT_SECTIONS: () => ELEMENT_SECTIONS,
  ElementType: () => ElementType,
  FIELD_SCHEMA: () => FIELD_SCHEMA,
  GameTier: () => GameTier,
  ONLYWORLDS_VERSION: () => ONLYWORLDS_VERSION,
  OnlyWorldsClient: () => OnlyWorldsClient,
  createAnyElementId: () => createAnyElementId,
  createElementId: () => createElementId,
  createElementIds: () => createElementIds,
  getElementIcon: () => getElementIcon,
  getElementLabel: () => getElementLabel,
  getElementSections: () => getElementSections
});
module.exports = __toCommonJS(index_exports);

// src/types.ts
var ElementType = /* @__PURE__ */ ((ElementType2) => {
  ElementType2["Ability"] = "ability";
  ElementType2["Character"] = "character";
  ElementType2["Collective"] = "collective";
  ElementType2["Construct"] = "construct";
  ElementType2["Creature"] = "creature";
  ElementType2["Event"] = "event";
  ElementType2["Family"] = "family";
  ElementType2["Institution"] = "institution";
  ElementType2["Language"] = "language";
  ElementType2["Law"] = "law";
  ElementType2["Location"] = "location";
  ElementType2["Map"] = "map";
  ElementType2["Marker"] = "marker";
  ElementType2["Narrative"] = "narrative";
  ElementType2["Object"] = "object";
  ElementType2["Phenomenon"] = "phenomenon";
  ElementType2["Pin"] = "pin";
  ElementType2["Relation"] = "relation";
  ElementType2["Species"] = "species";
  ElementType2["Title"] = "title";
  ElementType2["Trait"] = "trait";
  ElementType2["Zone"] = "zone";
  return ElementType2;
})(ElementType || {});
var ELEMENT_LABELS = {
  ["ability" /* Ability */]: "Abilities",
  ["character" /* Character */]: "Characters",
  ["collective" /* Collective */]: "Collectives",
  ["construct" /* Construct */]: "Constructs",
  ["creature" /* Creature */]: "Creatures",
  ["event" /* Event */]: "Events",
  ["family" /* Family */]: "Families",
  ["institution" /* Institution */]: "Institutions",
  ["language" /* Language */]: "Languages",
  ["law" /* Law */]: "Laws",
  ["location" /* Location */]: "Locations",
  ["map" /* Map */]: "Maps",
  ["marker" /* Marker */]: "Markers",
  ["narrative" /* Narrative */]: "Narratives",
  ["object" /* Object */]: "Objects",
  ["phenomenon" /* Phenomenon */]: "Phenomena",
  ["pin" /* Pin */]: "Pins",
  ["relation" /* Relation */]: "Relations",
  ["species" /* Species */]: "Species",
  ["title" /* Title */]: "Titles",
  ["trait" /* Trait */]: "Traits",
  ["zone" /* Zone */]: "Zones"
};
function getElementLabel(elementType) {
  return ELEMENT_LABELS[elementType];
}
var ELEMENT_SECTIONS = {
  ["ability" /* Ability */]: [
    { name: "Mechanics", order: 1, fields: ["activation", "duration", "potency", "range", "effects", "challenges", "talents", "requisites"] },
    { name: "World", order: 2, fields: ["prevalence", "tradition", "source", "locus", "instruments", "systems"] }
  ],
  ["character" /* Character */]: [
    { name: "Constitution", order: 1, fields: ["physicality", "mentality", "height", "weight", "species", "traits", "abilities"] },
    { name: "Origins", order: 2, fields: ["background", "motivations", "birth_date", "birthplace", "languages"] },
    { name: "World", order: 3, fields: ["reputation", "location", "objects", "institutions"] },
    { name: "Personality", order: 4, fields: ["charisma", "coercion", "competence", "compassion", "creativity", "courage"] },
    { name: "Social", order: 5, fields: ["family", "friends", "rivals"] },
    { name: "TTRPG", order: 6, fields: ["level", "hit_points", "STR", "DEX", "CON", "INT", "WIS", "CHA"] }
  ],
  ["collective" /* Collective */]: [
    { name: "Formation", order: 1, fields: ["composition", "count", "formation_date", "operator", "equipment"] },
    { name: "Dynamics", order: 2, fields: ["activity", "disposition", "state", "abilities", "symbolism"] },
    { name: "World", order: 3, fields: ["species", "characters", "creatures", "phenomena"] }
  ],
  ["construct" /* Construct */]: [
    { name: "Nature", order: 1, fields: ["rationale", "history", "status", "reach", "start_date", "end_date", "founder", "custodian"] },
    { name: "Involves", order: 2, fields: ["characters", "objects", "locations", "species", "creatures", "institutions", "traits", "collectives", "zones", "abilities", "phenomena", "languages", "families", "relations", "titles", "constructs", "events", "narratives"] }
  ],
  ["creature" /* Creature */]: [
    { name: "Biology", order: 1, fields: ["appearance", "weight", "height", "species"] },
    { name: "Behaviour", order: 2, fields: ["habits", "demeanor", "traits", "abilities", "languages"] },
    { name: "World", order: 3, fields: ["status", "birth_date", "location", "zone"] },
    { name: "TTRPG", order: 4, fields: ["challenge_rating", "hit_points", "armor_class", "speed", "actions"] }
  ],
  ["event" /* Event */]: [
    { name: "Nature", order: 1, fields: ["history", "challenges", "consequences", "start_date", "end_date", "triggers"] },
    { name: "Involves", order: 2, fields: ["characters", "objects", "locations", "species", "creatures", "institutions", "traits", "collectives", "zones", "abilities", "phenomena", "languages", "families", "relations", "titles", "constructs"] }
  ],
  ["family" /* Family */]: [
    { name: "Identity", order: 1, fields: ["spirit", "history", "traditions", "traits", "abilities", "languages", "ancestors"] },
    { name: "World", order: 2, fields: ["reputation", "estates", "governs", "heirlooms", "creatures"] }
  ],
  ["institution" /* Institution */]: [
    { name: "Foundation", order: 1, fields: ["doctrine", "founding_date", "parent_institution"] },
    { name: "Claims", order: 2, fields: ["zones", "objects", "creatures"] },
    { name: "World", order: 3, fields: ["status", "allies", "adversaries", "constructs"] }
  ],
  ["language" /* Language */]: [
    { name: "Structure", order: 1, fields: ["phonology", "grammar", "lexicon", "writing", "classification"] },
    { name: "World", order: 2, fields: ["status", "spread", "dialects"] }
  ],
  ["law" /* Law */]: [
    { name: "Code", order: 1, fields: ["declaration", "purpose", "date", "parent_law", "penalties"] },
    { name: "World", order: 2, fields: ["author", "locations", "zones", "prohibitions", "adjudicators", "enforcers"] }
  ],
  ["location" /* Location */]: [
    { name: "Setting", order: 1, fields: ["form", "function", "founding_date", "parent_location", "populations"] },
    { name: "Politics", order: 2, fields: ["political_climate", "primary_power", "governing_title", "secondary_powers", "zone", "rival", "partner"] },
    { name: "World", order: 3, fields: ["customs", "founders", "cults", "delicacies"] },
    { name: "Production", order: 4, fields: ["extraction_methods", "extraction_goods", "industry_methods", "industry_goods"] },
    { name: "Commerce", order: 5, fields: ["infrastructure", "extraction_markets", "industry_markets", "currencies"] },
    { name: "Construction", order: 6, fields: ["architecture", "buildings", "building_methods"] },
    { name: "Defense", order: 7, fields: ["defensibility", "elevation", "fighters", "defensive_objects"] }
  ],
  ["map" /* Map */]: [
    { name: "Details", order: 1, fields: ["background_color", "hierarchy", "width", "height", "depth", "parent_map", "location"] }
  ],
  ["marker" /* Marker */]: [
    { name: "Details", order: 1, fields: ["map", "zone", "x", "y", "z", "order"] }
  ],
  ["narrative" /* Narrative */]: [
    { name: "Context", order: 1, fields: ["story", "consequences", "start_date", "end_date", "order", "parent_narrative", "protagonist", "antagonist", "narrator", "conservator"] },
    { name: "Involves", order: 2, fields: ["events", "characters", "objects", "locations", "species", "creatures", "institutions", "traits", "collectives", "zones", "abilities", "phenomena", "languages", "families", "relations", "titles", "constructs", "laws"] }
  ],
  ["object" /* Object */]: [
    { name: "Form", order: 1, fields: ["aesthetics", "weight", "amount", "parent_object", "materials", "technology"] },
    { name: "Function", order: 2, fields: ["utility", "effects", "abilities", "consumes"] },
    { name: "World", order: 3, fields: ["origins", "location", "language", "affinities"] }
  ],
  ["phenomenon" /* Phenomenon */]: [
    { name: "Mechanics", order: 1, fields: ["expression", "effects", "duration", "catalysts", "empowerments"] },
    { name: "World", order: 2, fields: ["mythology", "system", "triggers", "wielders", "environments"] }
  ],
  ["pin" /* Pin */]: [
    { name: "Details", order: 1, fields: ["map", "element_type", "element_id", "element", "x", "y", "z"] }
  ],
  ["relation" /* Relation */]: [
    { name: "Nature", order: 1, fields: ["background", "start_date", "end_date", "intensity", "actor", "events"] },
    { name: "Involves", order: 2, fields: ["characters", "objects", "locations", "species", "creatures", "institutions", "traits", "collectives", "zones", "abilities", "phenomena", "languages", "families", "relations", "titles", "constructs", "events", "narratives"] }
  ],
  ["species" /* Species */]: [
    { name: "Biology", order: 1, fields: ["appearance", "life_span", "weight", "nourishment", "reproduction", "adaptations"] },
    { name: "Psychology", order: 2, fields: ["instincts", "sociality", "temperament", "communication", "aggression", "traits"] },
    { name: "World", order: 3, fields: ["role", "parent_species", "locations", "zones", "affinities"] }
  ],
  ["title" /* Title */]: [
    { name: "Mandate", order: 1, fields: ["authority", "eligibility", "grant_date", "revoke_date", "issuer", "body", "superior_title", "holders", "symbols"] },
    { name: "World", order: 2, fields: ["status", "history", "characters", "institutions", "families", "zones", "locations", "objects", "constructs", "laws", "collectives", "creatures", "phenomena", "species", "languages"] }
  ],
  ["trait" /* Trait */]: [
    { name: "Qualitative", order: 1, fields: ["social_effects", "physical_effects", "functional_effects", "personality_effects", "behaviour_effects"] },
    { name: "Quantitative", order: 2, fields: ["charisma", "coercion", "competence", "compassion", "creativity", "courage"] },
    { name: "World", order: 3, fields: ["significance", "anti_trait", "empowered_abilities"] }
  ],
  ["zone" /* Zone */]: [
    { name: "Scope", order: 1, fields: ["role", "start_date", "end_date", "phenomena", "linked_zones"] },
    { name: "World", order: 2, fields: ["context", "populations", "titles", "principles"] }
  ]
};
function getElementSections(elementType) {
  return ELEMENT_SECTIONS[elementType] || [];
}
var ONLYWORLDS_VERSION = "00.30.00";
var ELEMENT_ICONS = {
  ["ability" /* Ability */]: "auto_fix_normal",
  ["character" /* Character */]: "person",
  ["collective" /* Collective */]: "groups_3",
  ["construct" /* Construct */]: "api",
  ["creature" /* Creature */]: "bug_report",
  ["event" /* Event */]: "saved_search",
  ["family" /* Family */]: "supervisor_account",
  ["institution" /* Institution */]: "business",
  ["language" /* Language */]: "edit_road",
  ["law" /* Law */]: "gpp_bad",
  ["location" /* Location */]: "castle",
  ["map" /* Map */]: "map",
  ["marker" /* Marker */]: "place",
  ["narrative" /* Narrative */]: "menu_book",
  ["object" /* Object */]: "webhook",
  ["phenomenon" /* Phenomenon */]: "thunderstorm",
  ["pin" /* Pin */]: "push_pin",
  ["relation" /* Relation */]: "link",
  ["species" /* Species */]: "crib",
  ["title" /* Title */]: "military_tech",
  ["trait" /* Trait */]: "flaky",
  ["zone" /* Zone */]: "architecture"
};
var FIELD_SCHEMA = {
  ability: {
    // Base fields (shared by all elements)
    name: { type: "text", required: true },
    description: { type: "text", required: false },
    supertype: { type: "text", required: false },
    subtype: { type: "text", required: false },
    image_url: { type: "text", required: false },
    // Mechanics
    activation: { type: "text" },
    duration: { type: "number" },
    potency: { type: "number" },
    range: { type: "number" },
    effects: { type: "multi_link", target: "phenomenon" },
    challenges: { type: "text" },
    talents: { type: "multi_link", target: "trait" },
    requisites: { type: "multi_link", target: "construct" },
    // World
    prevalence: { type: "text" },
    tradition: { type: "single_link", target: "construct" },
    source: { type: "single_link", target: "phenomenon" },
    locus: { type: "single_link", target: "location" },
    instruments: { type: "multi_link", target: "object" },
    systems: { type: "multi_link", target: "construct" }
  },
  character: {
    // Base fields (shared by all elements)
    name: { type: "text", required: true },
    description: { type: "text", required: false },
    supertype: { type: "text", required: false },
    subtype: { type: "text", required: false },
    image_url: { type: "text", required: false },
    // Constitution
    physicality: { type: "text" },
    mentality: { type: "text" },
    height: { type: "number" },
    weight: { type: "number" },
    species: { type: "multi_link", target: "species" },
    traits: { type: "multi_link", target: "trait" },
    abilities: { type: "multi_link", target: "ability" },
    // Origins
    background: { type: "text" },
    motivations: { type: "text" },
    birth_date: { type: "number" },
    birthplace: { type: "single_link", target: "location" },
    languages: { type: "multi_link", target: "language" },
    // World
    reputation: { type: "text" },
    location: { type: "single_link", target: "location" },
    objects: { type: "multi_link", target: "object" },
    institutions: { type: "multi_link", target: "institution" },
    // Personality
    charisma: { type: "number" },
    coercion: { type: "number" },
    competence: { type: "number" },
    compassion: { type: "number" },
    creativity: { type: "number" },
    courage: { type: "number" },
    // Social
    family: { type: "multi_link", target: "family" },
    friends: { type: "multi_link", target: "character" },
    rivals: { type: "multi_link", target: "character" },
    // TTRPG
    level: { type: "number" },
    hit_points: { type: "number" },
    STR: { type: "number" },
    DEX: { type: "number" },
    CON: { type: "number" },
    INT: { type: "number" },
    WIS: { type: "number" },
    CHA: { type: "number" }
  },
  collective: {
    // Base fields (shared by all elements)
    name: { type: "text", required: true },
    description: { type: "text", required: false },
    supertype: { type: "text", required: false },
    subtype: { type: "text", required: false },
    image_url: { type: "text", required: false },
    // Formation
    composition: { type: "text" },
    count: { type: "number" },
    formation_date: { type: "number" },
    operator: { type: "single_link", target: "institution" },
    equipment: { type: "multi_link", target: "construct" },
    // Dynamics
    activity: { type: "text" },
    disposition: { type: "text" },
    state: { type: "text" },
    abilities: { type: "multi_link", target: "ability" },
    symbolism: { type: "multi_link", target: "construct" },
    // World
    species: { type: "multi_link", target: "species" },
    characters: { type: "multi_link", target: "character" },
    creatures: { type: "multi_link", target: "creature" },
    phenomena: { type: "multi_link", target: "phenomenon" }
  },
  construct: {
    // Base fields (shared by all elements)
    name: { type: "text", required: true },
    description: { type: "text", required: false },
    supertype: { type: "text", required: false },
    subtype: { type: "text", required: false },
    image_url: { type: "text", required: false },
    // Nature
    rationale: { type: "text" },
    history: { type: "text" },
    status: { type: "text" },
    reach: { type: "text" },
    start_date: { type: "number" },
    end_date: { type: "number" },
    founder: { type: "single_link", target: "character" },
    custodian: { type: "single_link", target: "institution" },
    // Involves
    characters: { type: "multi_link", target: "character" },
    objects: { type: "multi_link", target: "object" },
    locations: { type: "multi_link", target: "location" },
    species: { type: "multi_link", target: "species" },
    creatures: { type: "multi_link", target: "creature" },
    institutions: { type: "multi_link", target: "institution" },
    traits: { type: "multi_link", target: "trait" },
    collectives: { type: "multi_link", target: "collective" },
    zones: { type: "multi_link", target: "zone" },
    abilities: { type: "multi_link", target: "ability" },
    phenomena: { type: "multi_link", target: "phenomenon" },
    languages: { type: "multi_link", target: "language" },
    families: { type: "multi_link", target: "family" },
    relations: { type: "multi_link", target: "relation" },
    titles: { type: "multi_link", target: "title" },
    constructs: { type: "multi_link", target: "construct" },
    events: { type: "multi_link", target: "event" },
    narratives: { type: "multi_link", target: "narrative" }
  },
  creature: {
    // Base fields (shared by all elements)
    name: { type: "text", required: true },
    description: { type: "text", required: false },
    supertype: { type: "text", required: false },
    subtype: { type: "text", required: false },
    image_url: { type: "text", required: false },
    // Biology
    appearance: { type: "text" },
    weight: { type: "number" },
    height: { type: "number" },
    species: { type: "multi_link", target: "species" },
    // Behaviour
    habits: { type: "text" },
    demeanor: { type: "text" },
    traits: { type: "multi_link", target: "trait" },
    abilities: { type: "multi_link", target: "ability" },
    languages: { type: "multi_link", target: "language" },
    // World
    status: { type: "text" },
    birth_date: { type: "number" },
    location: { type: "single_link", target: "location" },
    zone: { type: "single_link", target: "zone" },
    // TTRPG
    challenge_rating: { type: "number" },
    hit_points: { type: "number" },
    armor_class: { type: "number" },
    speed: { type: "number" },
    actions: { type: "multi_link", target: "ability" }
  },
  event: {
    // Base fields (shared by all elements)
    name: { type: "text", required: true },
    description: { type: "text", required: false },
    supertype: { type: "text", required: false },
    subtype: { type: "text", required: false },
    image_url: { type: "text", required: false },
    // Nature
    history: { type: "text" },
    challenges: { type: "text" },
    consequences: { type: "text" },
    start_date: { type: "number" },
    end_date: { type: "number" },
    triggers: { type: "multi_link", target: "event" },
    // Involves
    characters: { type: "multi_link", target: "character" },
    objects: { type: "multi_link", target: "object" },
    locations: { type: "multi_link", target: "location" },
    species: { type: "multi_link", target: "species" },
    creatures: { type: "multi_link", target: "creature" },
    institutions: { type: "multi_link", target: "institution" },
    traits: { type: "multi_link", target: "trait" },
    collectives: { type: "multi_link", target: "collective" },
    zones: { type: "multi_link", target: "zone" },
    abilities: { type: "multi_link", target: "ability" },
    phenomena: { type: "multi_link", target: "phenomenon" },
    languages: { type: "multi_link", target: "language" },
    families: { type: "multi_link", target: "family" },
    relations: { type: "multi_link", target: "relation" },
    titles: { type: "multi_link", target: "title" },
    constructs: { type: "multi_link", target: "construct" }
  },
  family: {
    // Base fields (shared by all elements)
    name: { type: "text", required: true },
    description: { type: "text", required: false },
    supertype: { type: "text", required: false },
    subtype: { type: "text", required: false },
    image_url: { type: "text", required: false },
    // Identity
    spirit: { type: "text" },
    history: { type: "text" },
    traditions: { type: "multi_link", target: "construct" },
    traits: { type: "multi_link", target: "trait" },
    abilities: { type: "multi_link", target: "ability" },
    languages: { type: "multi_link", target: "language" },
    ancestors: { type: "multi_link", target: "character" },
    // World
    reputation: { type: "text" },
    estates: { type: "multi_link", target: "location" },
    governs: { type: "multi_link", target: "institution" },
    heirlooms: { type: "multi_link", target: "object" },
    creatures: { type: "multi_link", target: "creature" }
  },
  institution: {
    // Base fields (shared by all elements)
    name: { type: "text", required: true },
    description: { type: "text", required: false },
    supertype: { type: "text", required: false },
    subtype: { type: "text", required: false },
    image_url: { type: "text", required: false },
    // Foundation
    doctrine: { type: "text" },
    founding_date: { type: "number" },
    parent_institution: { type: "single_link", target: "institution" },
    // Claims
    zones: { type: "multi_link", target: "zone" },
    objects: { type: "multi_link", target: "object" },
    creatures: { type: "multi_link", target: "creature" },
    // World
    status: { type: "text" },
    allies: { type: "multi_link", target: "institution" },
    adversaries: { type: "multi_link", target: "institution" },
    constructs: { type: "multi_link", target: "construct" }
  },
  language: {
    // Base fields (shared by all elements)
    name: { type: "text", required: true },
    description: { type: "text", required: false },
    supertype: { type: "text", required: false },
    subtype: { type: "text", required: false },
    image_url: { type: "text", required: false },
    // Structure
    phonology: { type: "text" },
    grammar: { type: "text" },
    lexicon: { type: "text" },
    writing: { type: "text" },
    classification: { type: "single_link", target: "construct" },
    // World
    status: { type: "text" },
    spread: { type: "multi_link", target: "location" },
    dialects: { type: "multi_link", target: "language" }
  },
  law: {
    // Base fields (shared by all elements)
    name: { type: "text", required: true },
    description: { type: "text", required: false },
    supertype: { type: "text", required: false },
    subtype: { type: "text", required: false },
    image_url: { type: "text", required: false },
    // Code
    declaration: { type: "text" },
    purpose: { type: "text" },
    date: { type: "number" },
    parent_law: { type: "single_link", target: "law" },
    penalties: { type: "multi_link", target: "construct" },
    // World
    author: { type: "single_link", target: "institution" },
    locations: { type: "multi_link", target: "location" },
    zones: { type: "multi_link", target: "zone" },
    prohibitions: { type: "multi_link", target: "construct" },
    adjudicators: { type: "multi_link", target: "title" },
    enforcers: { type: "multi_link", target: "title" }
  },
  location: {
    // Base fields (shared by all elements)
    name: { type: "text", required: true },
    description: { type: "text", required: false },
    supertype: { type: "text", required: false },
    subtype: { type: "text", required: false },
    image_url: { type: "text", required: false },
    // Setting
    form: { type: "text" },
    function: { type: "text" },
    founding_date: { type: "number" },
    parent_location: { type: "single_link", target: "location" },
    populations: { type: "multi_link", target: "collective" },
    // Politics
    political_climate: { type: "text" },
    primary_power: { type: "single_link", target: "institution" },
    governing_title: { type: "single_link", target: "title" },
    secondary_powers: { type: "multi_link", target: "institution" },
    zone: { type: "single_link", target: "zone" },
    rival: { type: "single_link", target: "location" },
    partner: { type: "single_link", target: "location" },
    // World
    customs: { type: "text" },
    founders: { type: "multi_link", target: "character" },
    cults: { type: "multi_link", target: "construct" },
    delicacies: { type: "multi_link", target: "species" },
    // Production
    extraction_methods: { type: "multi_link", target: "construct" },
    extraction_goods: { type: "multi_link", target: "construct" },
    industry_methods: { type: "multi_link", target: "construct" },
    industry_goods: { type: "multi_link", target: "construct" },
    // Commerce
    infrastructure: { type: "text" },
    extraction_markets: { type: "multi_link", target: "location" },
    industry_markets: { type: "multi_link", target: "location" },
    currencies: { type: "multi_link", target: "construct" },
    // Construction
    architecture: { type: "text" },
    buildings: { type: "multi_link", target: "object" },
    building_methods: { type: "multi_link", target: "construct" },
    // Defense
    defensibility: { type: "text" },
    elevation: { type: "number" },
    fighters: { type: "multi_link", target: "construct" },
    defensive_objects: { type: "multi_link", target: "object" }
  },
  map: {
    // Base fields (shared by all elements)
    name: { type: "text", required: true },
    description: { type: "text", required: false },
    supertype: { type: "text", required: false },
    subtype: { type: "text", required: false },
    image_url: { type: "text", required: false },
    // Details
    background_color: { type: "text" },
    hierarchy: { type: "number" },
    width: { type: "number" },
    height: { type: "number" },
    depth: { type: "number" },
    parent_map: { type: "single_link", target: "map" },
    location: { type: "single_link", target: "location" }
  },
  marker: {
    // Base fields (shared by all elements)
    name: { type: "text", required: true },
    description: { type: "text", required: false },
    supertype: { type: "text", required: false },
    subtype: { type: "text", required: false },
    image_url: { type: "text", required: false },
    // Details
    map: { type: "single_link", target: "map" },
    zone: { type: "single_link", target: "zone" },
    x: { type: "number" },
    y: { type: "number" },
    z: { type: "number" },
    order: { type: "number" }
  },
  narrative: {
    // Base fields (shared by all elements)
    name: { type: "text", required: true },
    description: { type: "text", required: false },
    supertype: { type: "text", required: false },
    subtype: { type: "text", required: false },
    image_url: { type: "text", required: false },
    // Context
    story: { type: "text" },
    consequences: { type: "text" },
    start_date: { type: "number" },
    end_date: { type: "number" },
    order: { type: "number" },
    parent_narrative: { type: "single_link", target: "narrative" },
    protagonist: { type: "single_link", target: "character" },
    antagonist: { type: "single_link", target: "character" },
    narrator: { type: "single_link", target: "character" },
    conservator: { type: "single_link", target: "institution" },
    // Involves
    events: { type: "multi_link", target: "event" },
    characters: { type: "multi_link", target: "character" },
    objects: { type: "multi_link", target: "object" },
    locations: { type: "multi_link", target: "location" },
    species: { type: "multi_link", target: "species" },
    creatures: { type: "multi_link", target: "creature" },
    institutions: { type: "multi_link", target: "institution" },
    traits: { type: "multi_link", target: "trait" },
    collectives: { type: "multi_link", target: "collective" },
    zones: { type: "multi_link", target: "zone" },
    abilities: { type: "multi_link", target: "ability" },
    phenomena: { type: "multi_link", target: "phenomenon" },
    languages: { type: "multi_link", target: "language" },
    families: { type: "multi_link", target: "family" },
    relations: { type: "multi_link", target: "relation" },
    titles: { type: "multi_link", target: "title" },
    constructs: { type: "multi_link", target: "construct" },
    laws: { type: "multi_link", target: "law" }
  },
  object: {
    // Base fields (shared by all elements)
    name: { type: "text", required: true },
    description: { type: "text", required: false },
    supertype: { type: "text", required: false },
    subtype: { type: "text", required: false },
    image_url: { type: "text", required: false },
    // Form
    aesthetics: { type: "text" },
    weight: { type: "number" },
    amount: { type: "number" },
    parent_object: { type: "single_link", target: "object" },
    materials: { type: "multi_link", target: "construct" },
    technology: { type: "multi_link", target: "construct" },
    // Function
    utility: { type: "text" },
    effects: { type: "multi_link", target: "phenomenon" },
    abilities: { type: "multi_link", target: "ability" },
    consumes: { type: "multi_link", target: "construct" },
    // World
    origins: { type: "text" },
    location: { type: "single_link", target: "location" },
    language: { type: "single_link", target: "language" },
    affinities: { type: "multi_link", target: "trait" }
  },
  phenomenon: {
    // Base fields (shared by all elements)
    name: { type: "text", required: true },
    description: { type: "text", required: false },
    supertype: { type: "text", required: false },
    subtype: { type: "text", required: false },
    image_url: { type: "text", required: false },
    // Mechanics
    expression: { type: "text" },
    effects: { type: "text" },
    duration: { type: "number" },
    catalysts: { type: "multi_link", target: "object" },
    empowerments: { type: "multi_link", target: "ability" },
    // World
    mythology: { type: "text" },
    system: { type: "single_link", target: "phenomenon" },
    triggers: { type: "multi_link", target: "construct" },
    wielders: { type: "multi_link", target: "character" },
    environments: { type: "multi_link", target: "location" }
  },
  pin: {
    // Base fields (shared by all elements)
    name: { type: "text", required: true },
    description: { type: "text", required: false },
    supertype: { type: "text", required: false },
    subtype: { type: "text", required: false },
    image_url: { type: "text", required: false },
    // Details
    map: { type: "single_link", target: "map" },
    element_type: { type: "text" },
    // ElementType enum value
    element_id: { type: "single_link", target: "any" },
    // Can reference any element
    x: { type: "number" },
    y: { type: "number" },
    z: { type: "number" }
  },
  relation: {
    // Base fields (shared by all elements)
    name: { type: "text", required: true },
    description: { type: "text", required: false },
    supertype: { type: "text", required: false },
    subtype: { type: "text", required: false },
    image_url: { type: "text", required: false },
    // Nature
    background: { type: "text" },
    start_date: { type: "number" },
    end_date: { type: "number" },
    intensity: { type: "number" },
    actor: { type: "single_link", target: "character" },
    events: { type: "multi_link", target: "event" },
    // Involves
    characters: { type: "multi_link", target: "character" },
    objects: { type: "multi_link", target: "object" },
    locations: { type: "multi_link", target: "location" },
    species: { type: "multi_link", target: "species" },
    creatures: { type: "multi_link", target: "creature" },
    institutions: { type: "multi_link", target: "institution" },
    traits: { type: "multi_link", target: "trait" },
    collectives: { type: "multi_link", target: "collective" },
    zones: { type: "multi_link", target: "zone" },
    abilities: { type: "multi_link", target: "ability" },
    phenomena: { type: "multi_link", target: "phenomenon" },
    languages: { type: "multi_link", target: "language" },
    families: { type: "multi_link", target: "family" },
    relations: { type: "multi_link", target: "relation" },
    titles: { type: "multi_link", target: "title" },
    constructs: { type: "multi_link", target: "construct" },
    narratives: { type: "multi_link", target: "narrative" }
  },
  species: {
    // Base fields (shared by all elements)
    name: { type: "text", required: true },
    description: { type: "text", required: false },
    supertype: { type: "text", required: false },
    subtype: { type: "text", required: false },
    image_url: { type: "text", required: false },
    // Biology
    appearance: { type: "text" },
    life_span: { type: "number" },
    weight: { type: "number" },
    nourishment: { type: "multi_link", target: "species" },
    reproduction: { type: "multi_link", target: "construct" },
    adaptations: { type: "multi_link", target: "ability" },
    // Psychology
    instincts: { type: "text" },
    sociality: { type: "text" },
    temperament: { type: "text" },
    communication: { type: "text" },
    aggression: { type: "number" },
    traits: { type: "multi_link", target: "trait" },
    // World
    role: { type: "text" },
    parent_species: { type: "single_link", target: "species" },
    locations: { type: "multi_link", target: "location" },
    zones: { type: "multi_link", target: "zone" },
    affinities: { type: "multi_link", target: "phenomenon" }
  },
  title: {
    // Base fields (shared by all elements)
    name: { type: "text", required: true },
    description: { type: "text", required: false },
    supertype: { type: "text", required: false },
    subtype: { type: "text", required: false },
    image_url: { type: "text", required: false },
    // Mandate
    authority: { type: "text" },
    eligibility: { type: "text" },
    grant_date: { type: "number" },
    revoke_date: { type: "number" },
    issuer: { type: "single_link", target: "institution" },
    body: { type: "single_link", target: "institution" },
    superior_title: { type: "single_link", target: "title" },
    holders: { type: "multi_link", target: "character" },
    symbols: { type: "multi_link", target: "object" },
    // World
    status: { type: "text" },
    history: { type: "text" },
    characters: { type: "multi_link", target: "character" },
    institutions: { type: "multi_link", target: "institution" },
    families: { type: "multi_link", target: "family" },
    zones: { type: "multi_link", target: "zone" },
    locations: { type: "multi_link", target: "location" },
    objects: { type: "multi_link", target: "object" },
    constructs: { type: "multi_link", target: "construct" },
    laws: { type: "multi_link", target: "law" },
    collectives: { type: "multi_link", target: "collective" },
    creatures: { type: "multi_link", target: "creature" },
    phenomena: { type: "multi_link", target: "phenomenon" },
    species: { type: "multi_link", target: "species" },
    languages: { type: "multi_link", target: "language" }
  },
  trait: {
    // Base fields (shared by all elements)
    name: { type: "text", required: true },
    description: { type: "text", required: false },
    supertype: { type: "text", required: false },
    subtype: { type: "text", required: false },
    image_url: { type: "text", required: false },
    // Qualitative
    social_effects: { type: "text" },
    physical_effects: { type: "text" },
    functional_effects: { type: "text" },
    personality_effects: { type: "text" },
    behaviour_effects: { type: "text" },
    // Quantitative
    charisma: { type: "number" },
    coercion: { type: "number" },
    competence: { type: "number" },
    compassion: { type: "number" },
    creativity: { type: "number" },
    courage: { type: "number" },
    // World
    significance: { type: "text" },
    anti_trait: { type: "single_link", target: "trait" },
    empowered_abilities: { type: "multi_link", target: "ability" }
  },
  zone: {
    // Base fields (shared by all elements)
    name: { type: "text", required: true },
    description: { type: "text", required: false },
    supertype: { type: "text", required: false },
    subtype: { type: "text", required: false },
    image_url: { type: "text", required: false },
    // Scope
    role: { type: "text" },
    start_date: { type: "number" },
    end_date: { type: "number" },
    phenomena: { type: "multi_link", target: "phenomenon" },
    linked_zones: { type: "multi_link", target: "zone" },
    // World
    context: { type: "text" },
    populations: { type: "multi_link", target: "collective" },
    titles: { type: "multi_link", target: "title" },
    principles: { type: "multi_link", target: "construct" }
  }
};
function createElementId(id) {
  return id;
}
function createElementIds(ids) {
  return ids;
}
function createAnyElementId(id) {
  return id;
}

// src/token-resource.ts
var TokenResource = class {
  constructor(client) {
    this.client = client;
  }
  /**
   * Get current token status for authenticated user
   *
   * Returns daily token allowance, usage, and availability.
   * Matches base-tool's checkStatus() pattern.
   *
   * @returns Current token status
   * @example
   * ```typescript
   * const status = await client.tokens.getStatus();
   * console.log(`Available: ${status.tokens_available_today}/${status.token_rating}`);
   * console.log(`Used today: ${status.tokens_used_today}`);
   * console.log(`Active sessions: ${status.sessions_active}`);
   * ```
   */
  async getStatus() {
    return this.client.request("GET", "/tokens/status/");
  }
  /**
   * Consume tokens for service usage
   *
   * Reports token consumption to track daily usage. Allows consumption even if
   * exceeds available tokens (tracks as debt), but warns via error field.
   * Matches base-tool's reportUsage() pattern.
   *
   * @param params - Token consumption parameters
   * @returns Consumption result with updated balance
   * @example
   * ```typescript
   * const result = await client.tokens.consume({
   *   amount: 500,
   *   service: 'worldbuilding_tool',
   *   metadata: {
   *     feature: 'character_generation',
   *     model: 'gpt-4',
   *     prompt_tokens: 300,
   *     completion_tokens: 200
   *   }
   * });
   *
   * if (result.error) {
   *   console.warn('Token warning:', result.error);
   * }
   * console.log(`${result.tokens_remaining} tokens remaining`);
   * ```
   */
  async consume(params) {
    return this.client.request("POST", "/tokens/consume/", {
      body: {
        amount: params.amount,
        service: params.service || "sdk_client",
        session_id: params.sessionId ?? null,
        metadata: params.metadata ?? null
      }
    });
  }
  /**
   * Get encrypted OpenAI API key (advanced use case)
   *
   * Requires minimum 100 tokens available. Creates a 1-hour session for tracking.
   * Returns encrypted key that must be decrypted client-side using Fernet.
   *
   * See base-tool/src/llm/token-service.ts:99-155 for full implementation example
   * including client-side decryption with the 'fernet' npm package.
   *
   * @returns Encrypted access key and session info
   * @throws Error if insufficient tokens (< 100)
   * @example
   * ```typescript
   * // Get encrypted key
   * const access = await client.tokens.getAccessKey();
   *
   * // Decrypt using fernet library (see base-tool for full example)
   * // 1. Derive key from world ID using SHA-256
   * // 2. Use 'fernet' npm package to decrypt
   * // 3. Use decrypted OpenAI key for direct API calls
   * // 4. Report usage with access.session_id
   *
   * console.log('Session:', access.session_id);
   * console.log('Expires:', access.expires_at);
   * ```
   */
  async getAccessKey() {
    return this.client.request("GET", "/tokens/access-key/");
  }
  /**
   * Revoke a specific token session
   *
   * Invalidates the session ID obtained from getAccessKey().
   * Use when cleaning up or on logout.
   *
   * @param sessionId - Session ID to revoke
   * @returns Revocation result
   * @example
   * ```typescript
   * await client.tokens.revokeSession('session-id-here');
   * ```
   */
  async revokeSession(sessionId) {
    return this.client.request(
      "POST",
      `/tokens/revoke-session/?session_id=${encodeURIComponent(sessionId)}`
    );
  }
  /**
   * Revoke all active sessions (emergency use)
   *
   * Invalidates all token sessions for the authenticated user.
   * Use for security cleanup or when sessions are stuck.
   *
   * @returns Revocation result with count of revoked sessions
   * @example
   * ```typescript
   * const result = await client.tokens.revokeAllSessions();
   * console.log(`Revoked ${result.sessions_revoked} sessions`);
   * ```
   */
  async revokeAllSessions() {
    return this.client.request("POST", "/tokens/revoke-all-sessions/");
  }
  /**
   * Get public encryption info (no auth required)
   *
   * Returns algorithm details and example code for client-side decryption.
   * Public endpoint - can be called without authentication.
   *
   * @returns Encryption algorithm and implementation details
   * @example
   * ```typescript
   * const info = await client.tokens.getEncryptionInfo();
   * console.log('Algorithm:', info.algorithm);
   * console.log('Key derivation:', info.key_derivation);
   * console.log(info.javascript_example);
   * ```
   */
  async getEncryptionInfo() {
    return this.client.request("GET", "/tokens/encryption-info/");
  }
};

// src/client.ts
var Resource = class {
  constructor(client, elementType) {
    this.client = client;
    this.elementType = elementType;
  }
  async list(options) {
    const response = await this.client.request("GET", `/${this.elementType}/`, { params: options });
    if (Array.isArray(response)) {
      return {
        count: response.length,
        next: null,
        previous: null,
        results: response
      };
    }
    return response;
  }
  async get(id) {
    return this.client.request("GET", `/${this.elementType}/${id}/`);
  }
  async create(data) {
    let body = OnlyWorldsClient.prepareRelations(data, this.elementType);
    if (this.elementType === "pin") body = this.roundPinCoordinates(body);
    return this.client.request("POST", `/${this.elementType}/`, { body });
  }
  async update(id, data) {
    let body = OnlyWorldsClient.prepareRelations(data, this.elementType);
    if (this.elementType === "pin") body = this.roundPinCoordinates(body);
    return this.client.request("PATCH", `/${this.elementType}/${id}/`, { body });
  }
  async delete(id) {
    return this.client.request("DELETE", `/${this.elementType}/${id}/`);
  }
  /**
   * Round Pin coordinates to integers (API requirement)
   * @private
   */
  roundPinCoordinates(data) {
    const rounded = { ...data };
    if (typeof rounded.x === "number") rounded.x = Math.round(rounded.x);
    if (typeof rounded.y === "number") rounded.y = Math.round(rounded.y);
    if (typeof rounded.z === "number") rounded.z = Math.round(rounded.z);
    return rounded;
  }
};
var WorldResource = class {
  constructor(client) {
    this.client = client;
  }
  /**
   * Get the world associated with the current API key
   * Returns the world directly (not wrapped in pagination)
   */
  async get() {
    return this.client.request("GET", "/world/");
  }
  /**
   * Update the current world
   */
  async update(data) {
    return this.client.request("PATCH", "/world/", { body: data });
  }
};
var OnlyWorldsClient = class {
  constructor(config) {
    this.baseUrl = config.baseUrl || "https://www.onlyworlds.com/api/worldapi";
    this.headers = {
      "Content-Type": "application/json",
      "API-Key": config.apiKey,
      "API-Pin": config.apiPin
    };
    this.worlds = new WorldResource(this);
    this.tokens = new TokenResource(this);
    this.abilities = new Resource(this, "ability");
    this.characters = new Resource(this, "character");
    this.collectives = new Resource(this, "collective");
    this.constructs = new Resource(this, "construct");
    this.creatures = new Resource(this, "creature");
    this.events = new Resource(this, "event");
    this.families = new Resource(this, "family");
    this.institutions = new Resource(this, "institution");
    this.languages = new Resource(this, "language");
    this.laws = new Resource(this, "law");
    this.locations = new Resource(this, "location");
    this.maps = new Resource(this, "map");
    this.markers = new Resource(this, "marker");
    this.narratives = new Resource(this, "narrative");
    this.objects = new Resource(this, "object");
    this.phenomena = new Resource(this, "phenomenon");
    this.pins = new Resource(this, "pin");
    this.relations = new Resource(this, "relation");
    this.species = new Resource(this, "species");
    this.titles = new Resource(this, "title");
    this.traits = new Resource(this, "trait");
    this.zones = new Resource(this, "zone");
  }
  /**
   * Make a request to the OnlyWorlds API
   */
  async request(method, path, options) {
    const url = new URL(`${this.baseUrl}${path}`);
    if (options?.params) {
      Object.entries(options.params).forEach(([key, value]) => {
        if (value !== void 0 && value !== null) {
          url.searchParams.append(key, String(value));
        }
      });
    }
    const fetchOptions = {
      method,
      headers: this.headers
    };
    if (options?.body && ["POST", "PATCH", "PUT"].includes(method)) {
      fetchOptions.body = JSON.stringify(options.body);
    }
    const response = await fetch(url.toString(), fetchOptions);
    if (!response.ok) {
      let errorMessage = `API Error ${response.status}`;
      try {
        const errorText = await response.text();
        if (errorText) {
          try {
            const errorJson = JSON.parse(errorText);
            if (Array.isArray(errorJson.detail)) {
              const validationErrors = errorJson.detail.map((err) => {
                const location = err.loc ? err.loc.join(".") : "unknown";
                return `${location}: ${err.msg}`;
              }).join("; ");
              errorMessage += `: ${validationErrors}`;
            } else {
              errorMessage += `: ${errorJson.detail || errorJson.error || errorText}`;
            }
          } catch {
            errorMessage += `: ${errorText}`;
          }
        }
      } catch {
      }
      throw new Error(errorMessage);
    }
    if (response.status === 204) {
      return void 0;
    }
    return response.json();
  }
  /**
   * Helper to convert nested objects to _id/_ids format (legacy method)
   * @deprecated Use prepareRelations instead - it's called automatically in create/update
   */
  static prepareInput(data) {
    const result = { ...data };
    for (const [key, value] of Object.entries(result)) {
      if (value && typeof value === "object" && !Array.isArray(value) && "id" in value) {
        delete result[key];
        result[`${key}_id`] = value.id;
      } else if (Array.isArray(value) && value.length > 0 && typeof value[0] === "object" && "id" in value[0]) {
        delete result[key];
        result[`${key}_ids`] = value.map((item) => item.id);
      }
    }
    return result;
  }
  /**
   * Convert relation fields to API format (_id/_ids suffix)
   *
   * The OnlyWorlds API expects:
   * - single_link fields: fieldname_id (e.g., birthplace_id)
   * - multi_link fields: fieldname_ids (e.g., species_ids)
   *
   * This method auto-converts based on FIELD_SCHEMA:
   * - { species: ["id1", "id2"] } → { species_ids: ["id1", "id2"] }
   * - { birthplace: "location-id" } → { birthplace_id: "location-id" }
   * - { species: [{id: "id1", name: "X"}] } → { species_ids: ["id1"] }
   *
   * Called automatically by create() and update() methods.
   */
  static prepareRelations(data, elementType) {
    const schema = FIELD_SCHEMA[elementType];
    if (!schema) return data;
    const result = { ...data };
    for (const [key, value] of Object.entries(result)) {
      const fieldDef = schema[key];
      if (!fieldDef) continue;
      if (fieldDef.type === "single_link") {
        delete result[key];
        if (value && typeof value === "object" && "id" in value) {
          result[`${key}_id`] = value.id;
        } else if (typeof value === "string" || value === null) {
          result[`${key}_id`] = value;
        }
      } else if (fieldDef.type === "multi_link") {
        delete result[key];
        if (Array.isArray(value)) {
          if (value.length > 0 && typeof value[0] === "object" && "id" in value[0]) {
            result[`${key}_ids`] = value.map((item) => item.id);
          } else {
            result[`${key}_ids`] = value;
          }
        } else {
          result[`${key}_ids`] = [];
        }
      }
    }
    return result;
  }
};

// src/token-types.ts
var GameTier = /* @__PURE__ */ ((GameTier2) => {
  GameTier2["FREE"] = "free";
  GameTier2["SILVER"] = "silver";
  GameTier2["GOLD"] = "gold";
  GameTier2["PLATINUM"] = "platinum";
  GameTier2["DIAMOND"] = "diamond";
  GameTier2["DELUXE"] = "deluxe";
  return GameTier2;
})(GameTier || {});

// src/icon-utils.ts
var PLURAL_TO_SINGULAR = {
  abilities: "ability" /* Ability */,
  characters: "character" /* Character */,
  collectives: "collective" /* Collective */,
  constructs: "construct" /* Construct */,
  creatures: "creature" /* Creature */,
  events: "event" /* Event */,
  families: "family" /* Family */,
  institutions: "institution" /* Institution */,
  languages: "language" /* Language */,
  laws: "law" /* Law */,
  locations: "location" /* Location */,
  maps: "map" /* Map */,
  markers: "marker" /* Marker */,
  narratives: "narrative" /* Narrative */,
  objects: "object" /* Object */,
  phenomena: "phenomenon" /* Phenomenon */,
  // irregular plural
  pins: "pin" /* Pin */,
  relations: "relation" /* Relation */,
  species: "species" /* Species */,
  // same singular/plural
  titles: "title" /* Title */,
  traits: "trait" /* Trait */,
  zones: "zone" /* Zone */
};
function getElementIcon(type) {
  const lower = type.toLowerCase();
  if (lower in PLURAL_TO_SINGULAR) {
    return ELEMENT_ICONS[PLURAL_TO_SINGULAR[lower]];
  }
  const singular = Object.values(ElementType).find(
    (et) => et.toLowerCase() === lower
  );
  if (singular) {
    return ELEMENT_ICONS[singular];
  }
  return "help_outline";
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ELEMENT_ICONS,
  ELEMENT_LABELS,
  ELEMENT_SECTIONS,
  ElementType,
  FIELD_SCHEMA,
  GameTier,
  ONLYWORLDS_VERSION,
  OnlyWorldsClient,
  createAnyElementId,
  createElementId,
  createElementIds,
  getElementIcon,
  getElementLabel,
  getElementSections
});
